File: python/tests/integration/_sync/test_browser_context_integration.py
#!/usr/bin/env python3
"""
Integration test for browser context functionality.
This test verifies that browser context (cookies, localStorage, etc.) can be persisted
across sessions using the same ContextId.
"""

import asyncio
import os
import time

import pytest
from playwright.async_api import sync_playwright

from agentbay import AgentBay
from agentbay import BrowserContext, CreateSessionParams
from agentbay import BrowserOption


def get_test_api_key():
    """Get API key for testing"""
    api_key = os.environ.get("AGENTBAY_API_KEY")
    if not api_key:
        pytest.skip("AGENTBAY_API_KEY environment variable not set")
    return api_key


@pytest.fixture(scope="module")
def agent_bay():
    """Create AsyncAgentBay client for the test module."""
    # Skip if no API key is available or in CI environment
    api_key = os.environ.get("AGENTBAY_API_KEY")
    if not api_key or os.environ.get("CI"):
        pytest.skip("Skipping integration test: No API key available or running in CI")
    
    # Initialize AsyncAgentBay client
    return AgentBay(api_key)


@pytest.fixture(scope="module")
def browser_context_fixture(agent_bay):
    """Create and cleanup a browser context for the test module."""
    # Create a unique context name for this test
    context_name = f"test-browser-context-{int(time.time())}"
    
    # Create a context
    context_result = agent_bay.context.get(context_name, True)
    if not context_result.success or not context_result.context:
        pytest.skip("Failed to create context")
    
    context = context_result.context
    print(f"Created context: {context.name} (ID: {context.id})")
    
    yield context
    
    # Clean up context
    try:
        agent_bay.context.delete(context)
        print(f"Context deleted: {context.id}")
    except Exception as e:
        print(f"Warning: Failed to delete context: {e}")


@pytest.mark.asyncio
def test_browser_context_cookie_persistence(agent_bay, browser_context_fixture):
    """Test that manually set cookies persist across sessions with the same browser context."""
    context = browser_context_fixture
    
    # Test data
    test_url = "https://www.baidu.com"
    test_domain = "baidu.com"

    # Helper function to add one hour to current time
    def add_hour():
        return int(time.time()) + 3600  # 1 hour from now

    test_cookies = [
        {
            "name": "myCookie",
            "value": "cookieValue",
            "domain": test_domain,
            "path": "/",
            "httpOnly": False,
            "secure": False,
            "expires": add_hour(),
        },
        {
            "name": "test_cookie_2",
            "value": "test_value_2",
            "domain": test_domain,
            "path": "/",
            "httpOnly": False,
            "secure": False,
            "expires": add_hour(),
        },
    ]

    # Step 1 & 2: Create ContextId and create session with BrowserContext
    print(f"Step 1-2: Creating session with browser context ID: {context.id}")
    browser_context = BrowserContext(context.id, auto_upload=True)
    params = CreateSessionParams(
        image_id="browser_latest", browser_context=browser_context
    )

    session_result = agent_bay.create(params)
    if not session_result.success:
        print(f"Session creation failed!")
        print(f"Error message: {session_result.error_message}")
        print(f"Request ID: {session_result.request_id}")
    assert session_result.success, f"Failed to create first session: {session_result.error_message}"
    assert session_result.session is not None, "Session should not be None"

    session1 = session_result.session
    print(f"First session created with ID: {session1.session_id}")

    # Step 3: Get browser object through initialize and get_endpoint_url
    print("Step 3: Initializing browser and getting browser object...")

    # Initialize browser
    init_success = session1.browser.initialize(BrowserOption())
    assert init_success, "Failed to initialize browser"
    print("Browser initialized successfully")

    # Get endpoint URL
    endpoint_url = session1.browser.get_endpoint_url()
    assert endpoint_url is not None, "Endpoint URL should not be None"
    print(f"Browser endpoint URL: {endpoint_url}")

    # Step 4: Connect with playwright, open baidu.com and then add test cookies
    print("Step 4: Opening baidu.com and then adding test cookies...")
    with sync_playwright() as p:
        browser = p.chromium.connect_over_cdp(endpoint_url)
        assert browser is not None, "Failed to connect to browser"
        cdp_session = browser.new_browser_cdp_session()

        pw_context = (
            browser.contexts[0]
            if browser.contexts
            else browser.new_context()
        )
        page = pw_context.new_page()

        # Navigate to test URL first
        page.goto(test_url, timeout=60000)
        print(f"Navigated to {test_url}")

        # Wait a bit for the page to load
        page.wait_for_timeout(3000)

        # Add test cookies after navigating to the page
        pw_context.add_cookies(test_cookies)  # type: ignore
        print(
            f"Added {len(test_cookies)} test cookies after navigating to {test_url}"
        )

        # Read cookies to verify they were set correctly
        cookies = pw_context.cookies()
        cookie_dict = {
            cookie.get("name", ""): cookie.get("value", "")
            for cookie in cookies
        }

        print(f"Cookies found in first session: {list(cookie_dict.keys())}")
        print(f"Total cookies count: {len(cookies)}")

        # Store cookies for later verification
        first_session_cookies = cookies
        first_session_cookie_dict = cookie_dict

        cdp_session.send("Browser.close")
        print("First session browser operations completed")

        # Wait for browser to save cookies to file
        print("Waiting for browser to save cookies to file...")
        asyncio.sleep(2)
        print("Wait completed")

    # Step 5: Release session with syncContext=True
    print("Step 5: Releasing first session with syncContext=True...")
    delete_result = agent_bay.delete(session1, sync_context=True)
    assert delete_result.success, "Failed to delete first session"
    print(
        f"First session deleted successfully (RequestID: {delete_result.request_id})"
    )

    # Wait for context sync to complete
    asyncio.sleep(3)

    # Step 6: Create second session with same ContextId
    print(
        f"Step 6: Creating second session with same context ID: {context.id}"
    )
    session_result2 = agent_bay.create(params)
    assert session_result2.success, "Failed to create second session"
    assert session_result2.session is not None, "Second session should not be None"

    session2 = session_result2.session
    print(f"Second session created with ID: {session2.session_id}")

    # Step 7: Get browser object and check if test cookies exist without opening any page
    print(
        "Step 7: Getting browser object and checking test cookie persistence without opening any page..."
    )

    # Initialize browser
    init_success = session2.browser.initialize(BrowserOption())
    assert init_success, "Failed to initialize browser in second session"
    print("Second session browser initialized successfully")

    # Get endpoint URL
    endpoint_url = session2.browser.get_endpoint_url()
    assert endpoint_url is not None, "Endpoint URL should not be None"
    print(f"Second session browser endpoint URL: {endpoint_url}")

    # Connect with playwright and read cookies directly from context without opening any page
    with sync_playwright() as p:
        browser = p.chromium.connect_over_cdp(endpoint_url)
        assert browser is not None, "Failed to connect to browser in second session"

        pw_context = (
            browser.contexts[0]
            if browser.contexts
            else browser.new_context()
        )

        # Read cookies directly from context without opening any page
        cookies = pw_context.cookies()
        cookie_dict = {
            cookie.get("name", ""): cookie.get("value", "")
            for cookie in cookies
        }

        print(
            f"Cookies found in second session (without opening page): {list(cookie_dict.keys())}"
        )
        print(f"Total cookies count in second session: {len(cookies)}")

        # Check if our test cookies exist in the second session
        expected_cookie_names = {"myCookie", "test_cookie_2"}
        found_cookie_names = set(cookie_dict.keys())

        print(f"Expected test cookies: {expected_cookie_names}")
        print(f"Found cookies: {found_cookie_names}")

        # Check if all expected test cookies are present
        missing_cookies = expected_cookie_names - found_cookie_names
        assert not missing_cookies, (
            f"Missing expected test cookies in second session: {missing_cookies}"
        )

        # Check if test cookie values match what we set
        for cookie_name in expected_cookie_names:
            if cookie_name in cookie_dict:
                expected_value = next(
                    cookie["value"]
                    for cookie in test_cookies
                    if cookie["name"] == cookie_name
                )
                actual_value = cookie_dict[cookie_name]
                assert expected_value == actual_value, (
                    f"Test cookie '{cookie_name}' value should match. "
                    f"Expected: {expected_value}, Actual: {actual_value}"
                )
                print(
                    f"✓ Test cookie '{cookie_name}' value matches: {actual_value}"
                )

        print(
            f"SUCCESS: All {len(expected_cookie_names)} test cookies persisted correctly!"
        )
        print(f"Test cookies found: {list(expected_cookie_names)}")

        print("Step 8: Releasing second session with syncContext=True...")
        delete_result = agent_bay.delete(session2, sync_context=True)
        assert delete_result.success, "Failed to delete second session"
        print(
            f"Second session deleted successfully (RequestID: {delete_result.request_id})"
        )
        print("Second session browser operations completed")

    print("Browser context manual cookie persistence test completed successfully!")

File: java/agentbay/src/test/java/com/aliyun/agentbay/test/TestBrowserContextIntegration.java
package com.aliyun.agentbay.test;

import com.aliyun.agentbay.AgentBay;
import com.aliyun.agentbay.browser.BrowserOption;
import com.aliyun.agentbay.context.Context;
import com.aliyun.agentbay.context.ContextResult;
import com.aliyun.agentbay.exception.AgentBayException;
import com.aliyun.agentbay.model.DeleteResult;
import com.aliyun.agentbay.model.SessionResult;
import com.aliyun.agentbay.session.CreateSessionParams;
import com.aliyun.agentbay.session.Session;
import com.microsoft.playwright.*;
import com.microsoft.playwright.options.Cookie;
import org.junit.*;

import java.util.*;

import static org.junit.Assert.*;

/**
 * Integration test for browser context functionality.
 * This test verifies that browser context (cookies, localStorage, etc.) can be persisted
 * across sessions using the same ContextId.
 */
public class TestBrowserContextIntegration {
    
    private static AgentBay agentBay;
    private static String contextName;
    private static Context context;
    
    @BeforeClass
    public static void setUpClass() throws AgentBayException {
        // Skip if no API key is available or in CI environment
        String apiKey = System.getenv("AGENTBAY_API_KEY");
        String ci = System.getenv("CI");
        if (apiKey == null || apiKey.trim().isEmpty() || "true".equals(ci)) {
            throw new IllegalStateException(
                "Skipping integration test: No API key available or running in CI"
            );
        }
        
        // Initialize AgentBay client
        agentBay = new AgentBay();
        
        // Create a unique context name for this test
        contextName = "test-browser-context-" + System.currentTimeMillis();
        
        // Create a context
        ContextResult contextResult = agentBay.getContext().get(contextName, true);
        if (!contextResult.isSuccess() || contextResult.getContext() == null) {
            throw new IllegalStateException("Failed to create context");
        }
        
        context = contextResult.getContext();
        System.out.println("Created context: " + context.getName() + " (ID: " + context.getId() + ")");
    }
    
    @AfterClass
    public static void tearDownClass() {
        // Clean up context
        if (context != null && agentBay != null) {
            try {
                agentBay.getContext().delete(context);
                System.out.println("Context deleted: " + context.getId());
            } catch (Exception e) {
                System.out.println("Warning: Failed to delete context: " + e.getMessage());
            }
        }
    }
    
    /**
     * Test that manually set cookies persist across sessions with the same browser context.
     * 
     * This test verifies the complete browser context persistence workflow:
     * 1. Create a session with BrowserContext
     * 2. Initialize browser and add test cookies
     * 3. Delete session with context sync
     * 4. Create a new session with the same context
     * 5. Verify cookies are persisted
     */
    @Test
    public void testBrowserContextCookiePersistence() throws Exception {
        // Test data
        String testUrl = "https://www.aliyun.com";
        String testDomain = "aliyun.com";
        
        // Calculate expiration time (1 hour from now)
        double expirationTime = (System.currentTimeMillis() / 1000.0) + 3600;
        
        // Create test cookies using Playwright's Cookie builder
        List<Cookie> testCookies = new ArrayList<>();
        testCookies.add(new Cookie("myCookie", "cookieValue")
            .setDomain(testDomain)
            .setPath("/")
            .setExpires(expirationTime)
            .setHttpOnly(false)
            .setSecure(false));
        testCookies.add(new Cookie("test_cookie_2", "test_value_2")
            .setDomain(testDomain)
            .setPath("/")
            .setExpires(expirationTime)
            .setHttpOnly(false)
            .setSecure(false));
        
        // Step 1 & 2: Create ContextId and create session with BrowserContext
        System.out.println("Step 1-2: Creating session with browser context ID: " + context.getId());
        com.aliyun.agentbay.browser.BrowserContext browserContext = 
            new com.aliyun.agentbay.browser.BrowserContext(context.getId(), true);
        CreateSessionParams params = new CreateSessionParams();
        params.setImageId("browser_latest");
        params.setBrowserContext(browserContext);
        
        SessionResult sessionResult = agentBay.create(params);
        assertTrue("Failed to create first session", sessionResult.isSuccess());
        assertNotNull("Session should not be null", sessionResult.getSession());
        
        Session session1 = sessionResult.getSession();
        System.out.println("First session created with ID: " + session1.getSessionId());
        
        // Step 3: Get browser object through initialize and get_endpoint_url
        System.out.println("Step 3: Initializing browser and getting browser object...");
        
        // Initialize browser
        boolean initSuccess = session1.getBrowser().initialize(new BrowserOption());
        assertTrue("Failed to initialize browser", initSuccess);
        System.out.println("Browser initialized successfully");
        
        // Get endpoint URL
        String endpointUrl = session1.getBrowser().getEndpointUrl();
        assertNotNull("Endpoint URL should not be null", endpointUrl);
        System.out.println("Browser endpoint URL: " + endpointUrl);
        
        // Step 4: Connect with playwright, open aliyun.com and then add test cookies
        System.out.println("Step 4: Opening aliyun.com and then adding test cookies...");
        
        try (Playwright playwright = Playwright.create()) {
            Browser browser = playwright.chromium().connectOverCDP(endpointUrl);
            assertNotNull("Failed to connect to browser", browser);
            
            // Create CDP session for proper browser control
            CDPSession cdpSession = browser.newBrowserCDPSession();
            
            com.microsoft.playwright.BrowserContext pwContext = browser.contexts().isEmpty() 
                ? browser.newContext() 
                : browser.contexts().get(0);
            Page page = pwContext.newPage();
            
            // Navigate to test URL first
            page.navigate(testUrl);
            System.out.println("Navigated to " + testUrl);
            
            // Wait a bit for the page to load
            page.waitForTimeout(3000);
            
            // Add test cookies after navigating to the page
            pwContext.addCookies(testCookies);
            System.out.println("Added " + testCookies.size() + " test cookies after navigating to " + testUrl);
            
            // Read cookies to verify they were set correctly
            List<Cookie> cookies = pwContext.cookies();
            Map<String, String> cookieMap = new HashMap<>();
            for (Cookie cookie : cookies) {
                cookieMap.put(cookie.name, cookie.value);
            }
            
            System.out.println("Cookies found in first session: " + cookieMap.keySet());
            System.out.println("Total cookies count: " + cookies.size());
            
            // Close the browser using CDP session (same as Python version)
            cdpSession.send("Browser.close");
            System.out.println("First session browser operations completed");
            
            // Wait for browser to save cookies to file
            System.out.println("Waiting for browser to save cookies to file...");
            Thread.sleep(2000);
            System.out.println("Wait completed");
        }
        
        // Step 5: Release session with syncContext=true
        System.out.println("Step 5: Releasing first session with syncContext=true...");
        DeleteResult deleteResult = agentBay.delete(session1, true);
        assertTrue("Failed to delete first session", deleteResult.isSuccess());
        System.out.println("First session deleted successfully (RequestID: " + deleteResult.getRequestId() + ")");
        
        // Wait for context sync to complete
        Thread.sleep(3000);
        
        // Step 6: Create second session with same ContextId
        System.out.println("Step 6: Creating second session with same context ID: " + context.getId());
        SessionResult sessionResult2 = agentBay.create(params);
        assertTrue("Failed to create second session", sessionResult2.isSuccess());
        assertNotNull("Second session should not be null", sessionResult2.getSession());
        
        Session session2 = sessionResult2.getSession();
        System.out.println("Second session created with ID: " + session2.getSessionId());
        
        // Step 7: Get browser object and check if test cookies exist without opening any page
        System.out.println("Step 7: Getting browser object and checking test cookie persistence without opening any page...");
        
        // Initialize browser
        boolean initSuccess2 = session2.getBrowser().initialize(new BrowserOption());
        assertTrue("Failed to initialize browser in second session", initSuccess2);
        System.out.println("Second session browser initialized successfully");
        
        // Get endpoint URL
        String endpointUrl2 = session2.getBrowser().getEndpointUrl();
        assertNotNull("Endpoint URL should not be null", endpointUrl2);
        System.out.println("Second session browser endpoint URL: " + endpointUrl2);
        
        // Connect with playwright and read cookies directly from context without opening any page
        try (Playwright playwright = Playwright.create()) {
            Browser browser = playwright.chromium().connectOverCDP(endpointUrl2);
            assertNotNull("Failed to connect to browser in second session", browser);
            
            com.microsoft.playwright.BrowserContext pwContext = browser.contexts().isEmpty() 
                ? browser.newContext() 
                : browser.contexts().get(0);
            
            // Read cookies directly from context without opening any page
            List<Cookie> cookies = pwContext.cookies();
            Map<String, String> cookieMap = new HashMap<>();
            for (Cookie cookie : cookies) {
                cookieMap.put(cookie.name, cookie.value);
            }
            
            System.out.println("Cookies found in second session (without opening page): " + cookieMap.keySet());
            System.out.println("Total cookies count in second session: " + cookies.size());
            
            // Check if our test cookies exist in the second session
            Set<String> expectedCookieNames = new HashSet<>(Arrays.asList("myCookie", "test_cookie_2"));
            Set<String> foundCookieNames = cookieMap.keySet();
            
            System.out.println("Expected test cookies: " + expectedCookieNames);
            System.out.println("Found cookies: " + foundCookieNames);
            
            // Check if all expected test cookies are present
            Set<String> missingCookies = new HashSet<>(expectedCookieNames);
            missingCookies.removeAll(foundCookieNames);
            
            if (!missingCookies.isEmpty()) {
                fail("Missing expected test cookies in second session: " + missingCookies);
            }
            
            // Check if test cookie values match what we set
            Map<String, String> expectedValues = new HashMap<>();
            expectedValues.put("myCookie", "cookieValue");
            expectedValues.put("test_cookie_2", "test_value_2");
            
            for (String cookieName : expectedCookieNames) {
                if (cookieMap.containsKey(cookieName)) {
                    String expectedValue = expectedValues.get(cookieName);
                    String actualValue = cookieMap.get(cookieName);
                    assertEquals(
                        "Test cookie '" + cookieName + "' value should match. Expected: " + 
                        expectedValue + ", Actual: " + actualValue,
                        expectedValue,
                        actualValue
                    );
                    System.out.println("✓ Test cookie '" + cookieName + "' value matches: " + actualValue);
                }
            }
            
            System.out.println("SUCCESS: All " + expectedCookieNames.size() + " test cookies persisted correctly!");
            System.out.println("Test cookies found: " + expectedCookieNames);
            
            // Close context and browser
            pwContext.close();
            System.out.println("Second session browser operations completed");
        }
        
        // Clean up second session
        DeleteResult deleteResult2 = agentBay.delete(session2, false);
        assertTrue("Failed to delete second session", deleteResult2.isSuccess());
        System.out.println("Second session deleted successfully");
        
        System.out.println("Browser context manual cookie persistence test completed successfully!");
    }
    
    
    /**
     * Test that browser context can be created with minimal configuration
     */
    @Test
    public void testBrowserContextMinimalConfiguration() throws AgentBayException {
        System.out.println("Testing minimal browser context configuration...");
        
        // Create browser context with minimal config
        com.aliyun.agentbay.browser.BrowserContext browserContext = 
            new com.aliyun.agentbay.browser.BrowserContext(context.getId());
        
        // Verify default values
        assertEquals(context.getId(), browserContext.getContextId());
        assertTrue("Default auto_upload should be true", browserContext.isAutoUpload());
        
        // Create session with this context
        CreateSessionParams params = new CreateSessionParams();
        params.setImageId("browser_latest");
        params.setBrowserContext(browserContext);
        
        SessionResult sessionResult = agentBay.create(params);
        assertTrue("Failed to create session", sessionResult.isSuccess());
        assertNotNull("Session should not be null", sessionResult.getSession());
        
        Session session = sessionResult.getSession();
        System.out.println("Session created with minimal config: " + session.getSessionId());
        
        // Clean up
        DeleteResult deleteResult = agentBay.delete(session, false);
        assertTrue("Failed to delete session", deleteResult.isSuccess());
        System.out.println("Minimal configuration test completed successfully!");
    }
    
    /**
     * Test that browser context with auto_upload=false works correctly
     */
    @Test
    public void testBrowserContextNoAutoUpload() throws AgentBayException {
        System.out.println("Testing browser context with auto_upload=false...");
        
        // Create browser context with auto_upload=false
        com.aliyun.agentbay.browser.BrowserContext browserContext = 
            new com.aliyun.agentbay.browser.BrowserContext(context.getId(), false);
        
        // Verify configuration
        assertEquals(context.getId(), browserContext.getContextId());
        assertFalse("auto_upload should be false", browserContext.isAutoUpload());
        
        // Create session with this context
        CreateSessionParams params = new CreateSessionParams();
        params.setImageId("browser_latest");
        params.setBrowserContext(browserContext);
        
        SessionResult sessionResult = agentBay.create(params);
        assertTrue("Failed to create session", sessionResult.isSuccess());
        assertNotNull("Session should not be null", sessionResult.getSession());
        
        Session session = sessionResult.getSession();
        System.out.println("Session created with auto_upload=false: " + session.getSessionId());
        
        // Clean up
        DeleteResult deleteResult = agentBay.delete(session, false);
        assertTrue("Failed to delete session", deleteResult.isSuccess());
        System.out.println("No auto-upload test completed successfully!");
    }
}

